### PR TITLE
wip: Allocator interface proposal.

### PR DESCRIPTION
**Pull request by NicolasDenoyelle**
This PR is meant as an iterative discussion thread on allocator design and implementation.
This first commit introduces the high level interface header.

### CODE REVIEW

**Review by perarnau — COMMENTED** [review]
squash these 3 commits together

File: benchmarks/utils.c [hunk 1]
```diff
@@ -16,3 +20,54 @@ long long int aml_timediff(struct timespec start, struct timespec end)
...
+	return AML_SUCCESS;
+}
+
+int aml_time_stats_add(struct aml_time_stats *out, const long long val)
+{
+	if (out == NULL)
+		return -AML_EINVAL;
+
+	out->n++;
+	out->mean += ((double)val - out->mean) / (double)out->n;
```
Comment (perarnau) [inline benchmarks/utils.c:43]: just sum the values directly, and compute the mean at the end, that's easier

File: src/allocator/allocator.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+void *aml_aligned_alloc(struct aml_allocator *allocator,
+                        size_t size,
+                        size_t alignement)
+{
+	if (allocator == NULL || allocator->data == NULL ||
+	    allocator->ops == NULL || allocator->ops->alloc == NULL) {
+		aml_errno = AML_EINVAL;
+		return NULL;
+	}
+
```
Comment (perarnau) [inline src/allocator/allocator.c:35]: check for power of two

File: include/aml/higher/allocator.h [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+void *aml_alloc(struct aml_allocator *allocator, size_t size);
+
+/**
+ * Allocation routine with a pointer aligned on a specific boundary.
+ * Allocation may not be aligned if the allocator does not support aligned
+ * allocations.
+ *
+ * @param[in, out] allocator: The allocator to use.
+ * @param[in] size: The minimum allocation size.
+ * @param[in] alignement: The allocation alignement.
```
Comment (perarnau) [inline include/aml/higher/allocator.h:143]: maybe be more specific about the nature of the alignment, number of bytes or number of bits at 0.

File: include/aml/higher/allocator/area.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ */
+int aml_allocator_area_create(struct aml_allocator **out,
+                              struct aml_area *area);
+
+/** Destroy an area allocator. The embedded area is not destroyed. */
+int aml_allocator_area_destroy(struct aml_allocator **allocator);
+
+/**
+ * `alloc()` method of the allocator.
+ * Passes `NULL` to `aml_area_mmap()` options argument.
```
Comment (perarnau) [inline include/aml/higher/allocator/area.h:55]: add pointer in allocator_data to corresponding area mmap options.

File: include/aml/higher/allocator/utils.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                              const struct aml_memory_chunk sub);
+
+struct aml_vector;
+
+/** One pool of memory. */
+struct aml_memory_pool {
+	// The chunk of memory that this pool manages.
+	struct aml_memory_chunk memory;
+	// The chunks of memory still in the pool.
+	struct aml_vector *chunks;
```
Comment (perarnau) [inline include/aml/higher/allocator/utils.h:84]: improve comments here: memory is the total space allocatable, the vector contains chunks inside this space

File: src/allocator/utils.c [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+int aml_memory_chunk_match(struct aml_memory_chunk a, struct aml_memory_chunk b)
+{
+	return (size_t)a.ptr + a.size == (size_t)b.ptr ||
+	       (size_t)b.ptr + b.size == (size_t)a.ptr;
+}
+
+int aml_memory_chunk_overlap(struct aml_memory_chunk a,
+                             struct aml_memory_chunk b)
+{
+	if ((size_t)a.ptr <= (size_t)b.ptr &&
```
Comment (perarnau) [inline src/allocator/utils.c:41]: uintptr_t and the diff between two pointers might be cleaner

File: benchmarks/Makefile.am [hunk 7]
```diff
@@ -52,3 +57,10 @@ TESTS = $(BENCHMARKS)
 # phony target to allow us to compile the check programs without running the
 # tests. Required for any environment where we are cross-compiling
 check-programs-local: $(BENCHMARKS)
+
+# Rule to clean files generated by benchmarks.
+# https://www.gnu.org/software/automake/manual/html_node/Clean.html
+clean-local: clean-local-check
```
Comment (perarnau) [inline benchmarks/Makefile.am:67]: check if the log files generated by the automake infra are enough, .trs and .log are generated on make check

**Review by Kerilk — APPROVED** [review]

**Pull request merged**